pkg/fqdn/dnsproxy/proxy.go

```diff
@@ -268,52 +338,151 @@ func (p *DNSProxy) RemoveRestoredRules(endpointID uint16) {
+	fqdnKey := strings.Join(fqdnList, "|")
+	if entry, ok := allow.nameMatcherBySignature[fqdnKey]; ok {
+		entry.referenceCount += 1
+		return entry.matcher
+	}
+	f := &matchNameMatcher{
+		mapping: make(map[string]struct{}, len(fqdnList)),
+		key:     &fqdnKey,
+	}
+	for _, fqdn := range fqdnList {
+		f.mapping[fqdn] = struct{}{}
+	}
+	allow.nameMatcherBySignature[fqdnKey] = &matchNamesEntry{matcher: f, referenceCount: 1}
+	return f
+}
+
+// releaseNameMatcher releases the name matcher. In case there are no longer any references to
+// it, it will be freed.
+func (allow perEPAllow) releaseNameMatcher(entry *matchNameMatcher) {
+	if indexEntry, ok := allow.nameMatcherBySignature[*entry.key]; ok {
+		switch indexEntry.referenceCount {
+		case 1:
+			delete(allow.nameMatcherBySignature, *entry.key)
+			for k := range entry.mapping {
+				delete(entry.mapping, k)
+			}
+		default:
+			indexEntry.referenceCount -= 1
+		}
+	}
+}
+
+// removeAndReleasePortRulesForID removes the old port rules for the given destPort on the given endpointID. It also
+// releases the matchers so that unused matchers can be freed from memory.
+func (allow perEPAllow) removeAndReleasePortRulesForID(endpointID uint64, destPort uint16) {
+	epPorts, hasEpPorts := allow.allowMap[endpointID]
+	if !hasEpPorts {
+		return
+	}
+	for _, e := range epPorts[destPort] {
+		allow.releasePatternMatcher(e.patternMatcher)
+		allow.releaseNameMatcher(e.nameMatcher)
+	}
+	delete(epPorts, destPort)
+	if len(epPorts) == 0 {
+		delete(allow.allowMap, endpointID)
+	}
+}
+
 // setPortRulesForID sets the matching rules for endpointID and destPort for
-// later lookups. It converts newRules into a unified regexp that can be reused
-// later.
+// later lookups. It converts newRules into a regex for the matchPatterns and a lookup map for the matchNames.
 func (allow perEPAllow) setPortRulesForID(endpointID uint64, destPort uint16, newRules policy.L7DataMap) error {
-	// This is the delete case
 	if len(newRules) == 0 {
-		epPorts := allow[endpointID]
-		delete(epPorts, destPort)
-		if len(epPorts) == 0 {
-			delete(allow, endpointID)
-		}
+		allow.removeAndReleasePortRulesForID(endpointID, destPort)
 		return nil
 	}
+	cse := make(CachedSelectorEntry, len(newRules))
+	var err error
+	for selector, newRuleset := range newRules {
+		regexPattern, fqdnsList := GenerateRegexpAndFqdns(newRuleset)
+
+		regex, err := allow.lookupOrCompilePatternMatcher(regexPattern)
+		if err != nil {
+			break
+		}
 
-	newRE, err := GetSelectorRegexMap(newRules)
+		fqdnMapping := allow.lookupOrCreateNameMatcher(fqdnsList)
+		cse[selector] = fqdnMatcher{
+			patternMatcher: regex,
+			nameMatcher:    fqdnMapping,
+		}
+	}
 	if err != nil {
+		// Unregister all the registered regexps and fqdn mappings before exiting
+		for k, ruleset := range cse {
+			allow.releasePatternMatcher(ruleset.patternMatcher)
+			allow.releaseNameMatcher(ruleset.nameMatcher)
```

Comment:
Might make sense to have a function that handles both of the releases altogether so that someone in the future doesn't potentially forget to call one or the other.

Reply:
Yee, smart! Will at a separate func for that!